### ISSUE TITLE
Plan comment on PR?

### ISSUE THREAD

**Issue by ZebraFlesh** [hashicorp/setup-terraform#3]
Since the old action is deprecated, how can I make the new action put a comment with output of `terraform plan` on my PR?

**Comment by ZebraFlesh** [comment]
Turns out it's in the readme, but a bit buried: https://github.com/hashicorp/setup-terraform/blame/master/README.md#L96-L104

Might be worth elevating this as a separate topic in the readme.

**Comment by ericmatte** [comment]
Yeah, but with this method, you would have to add a `uses: actions/github-script@0.9.0` step for each terraform commands.

With the old action, each step that I used (`init`, `validate` and `plan`) could output a comment on Github.

**Comment by imjohnbo** [comment]
> Yeah, but with this method, you would have to add a uses: actions/github-script@0.9.0 step for each terraform command.

Or you could pipe the output into an action you find in Marketplace, maybe one that creates this pull request comment, makes an issue, sends a Slack notification...whatever you need!

**Comment by sudomateo** [comment]
Removing the ability to comment on pull requests from the action was an intentional design decision we made when creating this new action. Instead of forcing an opinionated stance on how output should be formatted or where output should be posted to we chose to implement an opt-out wrapper script that would expose the STDOUT, STDERR, and exit code of a given execution of `terraform`. This allows a user to transform or otherwise use the output any way they'd like. This also also this action to focus on doing one thing well which is executing `terraform`. There was a lot of discussion about this internally since commenting on pull requests was such a highly desired feature of the old action and simultaneously its largest pain point. Although the workflow YAML file will now require more steps if you wish to use the output of multiple `terraform` commands, the ability to have more control over the output should be worth the trade off.

**Comment by ZebraFlesh** [comment]
I guess I don't understand the point of this action then: terraform is already installed on the GitHub hosted runners, so if this action can't comment on a PR ... why have it?

After playing with this for a bit, the proof of concept PR commenting in the readme is woefully inadequate as compared to the old action. Things I've fought with, failed, and am moving on from because they're not worth my time (but I was this action would just handle for me):
* determining if `terraform plan` was successful with changes vs. successful without changes
* successful/failure comment
* details/summary comment to capture details for reviewers, but keep them out of the way

These are significant functionality gaps between this new action and the old one.

**Comment by sudomateo** [comment]
I appreciate your feedback! Here's some more information that might help show why these decisions were made.

> I guess I don't understand the point of this action then: terraform is already installed on the GitHub hosted runners,

The version of Terraform installed in the runners is consistently updated. This action allows a user to specify their own version of Terraform to install in addition to configuring the CLI configuration file and installing a wrapper script. This action is meant to set up Terraform for subsequent steps to consume much like the other `setup-*` GitHub actions. It's possible to forgo this action, or any `setup-*` GitHub action really, and perform all of the operations yourself. Doing so would create a more complex workflow file, something that was already a pain point of the old action.

> so if this action can't comment on a PR ... why have it?

To allow users to install Terraform, configure the CLI configuration file, and gain access to STDOUT, STDERR, and the exit code for each subsequent call to `terraform`. The previous action made some assumptions in order to provide the pull request commenting that you are familiar with. These assumptions led to some limitations. For example, the previous action had to wrap each individual Terraform subcommand in order to grab its output and then make a decision about how to format it. This led to the limitation of not allowing the user to use certain Terraform subcommands like `workspace`, `show`, etc. Also, the previous action was only able to comment on GitHub pull requests. Now, the output is exposed to the user for them to do with it what they want.

> These are significant functionality gaps between this new action and the old one.

The functionality gaps you refer to here are regarding pull request comments. While I can agree that commenting functionality is not 100% equivalent to the previous action (How can it be when this action does not perform the commenting?), it leaves room for other actions to be developed to meet these use cases later on and not add bloat to the `hashicorp/setup-terraform` action. The old action was attempting to do too much in one action, which is one of the reasons we decided to simplify this `hashicorp/setup-terraform` action. The ability for this action to comment on pull requests was intentionally left out, but not forgotten, in order to provide the other benefits of Terraform such as free-form arguments, the ability to access all of Terraform's subcommands, and the ability for Terraform to access third-party tools during the run (`aws`, `~/.ssh/*`, `ansible`, etc). As GitHub Actions matures, we expect there will be opportunity to add more functionality to how Terraform can be utilized in a workflow.

**Comment by jalaziz** [comment]
Another issue with the plan commenting is you lose the coloring in the action logs if you don't want messed up output on the comment. This probably could be done by introducing another action to filter the colors from stdout output (or maybe it can be included in the wrapper?).

While I'm sure this is probably better in the long run, it definitely seems more burdensome for getting up and running quickly.

**Comment by GABeech** [comment]
>  it leaves room for other actions to be developed to meet these use cases later on and not add bloat to the hashicorp/setup-terraform action

If you had not deprecated the old action, but simply moved the setup piece here ... that would have made a ton of sense. But, now you are doing that, and committing to maintaining a wrapper script _anyway_ why not just maintain both actions? This one is if you need to configure more than the defaults, the other one to actually do the work? That's how most other systems with `setup-*` action work. The setup dos the global config, so you don't have to re-specify variables every time, and the other action does the work.

**Comment by sudomateo** [comment]
> That's how most other systems with setup-* action work. The setup dos the global config, so you don't have to re-specify variables every time, and the other action does the work.

This _was_ the initial plan. We went in the current direction due to time, resource, and prioritization constraints. It is not out of the picture to go this route in the future.
